### ISSUE TITLE
BUG: special: Unchecked calls of malloc() and calloc() in specfun.h

### ISSUE THREAD

**Issue by WarrenWeckesser** [scipy/scipy#21417]
### Describe your issue.

Scattered throughout https://github.com/scipy/scipy/blob/main/scipy/special/xsf/specfun/specfun.h are calls of `malloc()` and `calloc()` where the return values are used without being checked for failure.  That will cause a seg. fault if an allocation does ever fail, which is not correct behavior.  Failures of memory allocation must ultimately raise a `MemoryError`.

We did a lot of work a few years ago to clean up such issues (https://github.com/scipy/scipy/issues/9999).  Perhaps it is time for another audit.

**Comment by ilayn** [comment]
That's my laziness when I was converting fortran stack array allocations. I'm not sure whether conversion to heap was a good move but pushing 1000 items to stack is also not good either.

**Comment by WarrenWeckesser** [comment]
> That's my laziness when I was converting fortran stack array allocations.

Yeah, I understand!  But eventually the technical debt must be paid.

It looks like this occurs in C++ functions with `void` return type.  They could be changed to return an `int` status code (eg. 0 means OK, -1 means memory allocation failed). (Or, as @steppi mentioned above, a pointer to a status code could be added as a parameter.)  The wrappers must check this and propagate the error code up the wrapping layers until the top level wrapper can detect it and set a Python `MemoryError` exception.

**Comment by ilayn** [comment]
In general, it does not feell right to alloc for a special function if it is running over a long array. 

Especially functions like `kmn` doing some expansion coefficients, every time, not sure that's the right thing to do. But I don't know enough special to comment except the code was subpar at certain places.

**Comment by steppi** [comment]
> It looks like this occurs in C++ functions with `void` return type. They could be changed to return an `int` status code (eg. 0 means OK, -1 means memory allocation failed). (Or, as @steppi mentioned above, a pointer to a status code could be added as a parameter.) The wrappers must check this and propagate the error code up the wrapping layers until the top level wrapper can detect it and set a Python `MemoryError` exception.

I see. I didn't realize they all have `void` return type. Yes, we can just change them to return an `int` status code. Right now memory errors are getting lumped into `SF_ERROR_OTHER` in `special`'s error handling,

e.g.

```C++
    if (eg == NULL) {
        set_error("prolate_segv", SF_ERROR_OTHER, "memory allocation error");
        return std::numeric_limits<T>::quiet_NaN();
    }
```

 but we could add `SF_ERROR_MEMORY` to `sf_error.cc` and `sf_error_state.c`. For a first pass, I think it's better to keep these as `SF_ERROR_OTHER` though, and then add a new error state in a follow up.

**Comment by steppi** [comment]
> In general, it does not feell right to alloc for a special function/ufunc if it is running over a long array.
> 
> Especially functions like `kmn` doing some expansion coefficients, every time, not sure that's the right thing to do. But I don't know enough special to comment except the code was subpar at certain places.

It’s tough to avoid when the expansion coefficients depend on the parameters. The functions with these allocs won’t work on GPU though so if possible, I want to find an alternative code path that gets switched in with macros when on CUDA. When the expansion coefficients are only used once, they don’t need to be stored, and when they are used multiple times, it might be efficient to just recompute them and make up for the redundant computation with parallelism. It would take some difficult refactoring to do this though.

**Comment by izaid** [comment]
> > It looks like this occurs in C++ functions with `void` return type. They could be changed to return an `int` status code (eg. 0 means OK, -1 means memory allocation failed). (Or, as @steppi mentioned above, a pointer to a status code could be added as a parameter.) The wrappers must check this and propagate the error code up the wrapping layers until the top level wrapper can detect it and set a Python `MemoryError` exception.
> 
> I see. I didn't realize they all have `void` return type. Yes, we can just change them to return an `int` status code. Right now memory errors are getting lumped into `SF_ERROR_OTHER` in `special`'s error handling,
> 
> e.g.
> 
> ```c++
>     if (eg == NULL) {
>         set_error("prolate_segv", SF_ERROR_OTHER, "memory allocation error");
>         return std::numeric_limits<T>::quiet_NaN();
>     }
> ```
> 
> but we could add `SF_ERROR_MEMORY` to `sf_error.cc` and `sf_error_state.c`. For a first pass, I think it's better to keep these as `SF_ERROR_OTHER` though, and then add a new error state in a follow up.

Chiming in here. We shouldn't change the `void` return type to return an array code, as that actually messes up the direct mapping to ufuncs that we recently put a lot of time into. I actually removed a lot of the old behaviour in which functions returned an error code. I think @steppi's idiom was right:

```
T* u = new (std::nothrow) T[nn + 4];
if (!u) {
    set_error(func_name, SF_ERROR_OTHER, "memory allocation error");
    return std::numeric_limits<T>::quiet_NaN();
}
```

@WarrenWeckesser I see you've been doing various things with special recently, would you also be able to cc me (in addition to @steppi) when you make these issues / PRs? Otherwise I don't necessarily see them. I did a lot of work in `special` recently and there were reasons for some of these changes.

**Comment by izaid** [comment]
About the `malloc` / `calloc` issue in general, I saw it and at some point wanted to address it. Memory obviously should be on the stack if possible, but if not we should just use the error reporting mechanism in `special` that already exists (as above). If we have to return something alongside that, then it should just be `NaN`.

**Comment by steppi** [comment]
> > 
> 
> Chiming in here. We shouldn't change the `void` return type to return an errir code, as that actually messes up the direct mapping to ufuncs that we recently put a lot of time into. I actually removed a lot of the old behaviour in which functions returned an error code. I think @steppi's idiom was right:

Hey Irwin. Sorry for not pinging you on this. Are you sure having a non `void` return type wouldn't be kosher in this case? We're only talking about internal functions which take pointers or references to arrays and modify them in some way, usually for things like generating expansion coefficients that are input dependent. Nothing of this nature would be used as a kernel for a ufunc.  Nevertheless, an argument I see in favor of still returning void and passing a pointer to where a status code can be set, is that some internal functions actually return values, and it may be nice to have a consistent way to get a status code from an internal function which works regardless of the return type.

**Comment by ilayn** [comment]
If it is possible we should not deal with errors, since I did not have ufuncs in mind I did a casual choice but now it is not a discussion. Stack or heap this memory should not be allocated 1000 times if it is running over a 1000-long array. I don't know if it is possible now but the proper way is to allocate the memory outside and at the start of ufunc and reuse it for every invocation of the function running over the array. That would resolve both issues at the same time. But I can't decipher if this is possible in the current ufunc machinery.

**Comment by steppi** [comment]
> If it is possible we should not deal with errors, since I did not have ufuncs in mind I did a casual choice but now it is not a discussion. Stack or heap this memory should not be allocated 1000 times if it is running over a 1000-long array. I don't know if it is possible now but the proper way is to allocate the memory outside and at the start of ufunc and reuse it for every invocation of the function running over the array. That would resolve both issues at the same time. But I can't decipher if this is possible in the current ufunc machinery.

Yeah, when the contents of an array aren’t dependent on the functions inputs, the convention I think should be that they are `constexpr` arrays defined outside the functions. I think there are some cases like that left that are currently defined inside functions that we can change. The main problem we’re dealing with though is difficult to compute functions, where the size and content of the arrays depend on the functions inputs. That’s usually what’s happening where we now have all of those mallocs in specfun. It’s not a given that there will be redundant computation and memory use when computing over an array. I’ve thought about implementing some kind of cached ufunc before though, with caches that persist just over one Python level call of a ufunc, where you can mark intermediate functions used in a scalar kernel as cacheable. I think that would be a pretty useful optimization, because the expansion coefficients depend on just some of arguments, often just one in a multiarg function.

**Comment by steppi** [comment]
> In most cases it over-allocates that's why it is asking 100 a fixed number. But uses input dependent many entries.

Ah, that’s right. Thanks for pointing that out again. We can change those to only allocate what’s needed.

**Comment by izaid** [comment]
I just think we shouldn't have two different kinds of error reporting in `xsf`. We've got the `set_error` system, which we inherited and ugpraded from what was in SciPy before. We should use that instead of returning an error code, unless we decide we want to return error codes instead of use that. Any error code ultimately needs to be translated to something else anyway.

**Comment by steppi** [comment]
> I just think we shouldn't have two different kinds of error reporting in `xsf`. We've got the `set_error` system, which we inherited and ugpraded from what was in SciPy before. We should use that instead of returning an error code, unless we decide we want to return error codes instead of use that. Any error code ultimately needs to be translated to something else anyway.

I think we’re talking past each other. The issue is that if there are internal functions and tools which are used in multiple ufunc scalar kernels, then we won’t know the name of the calling ufunc if something goes wrong in the internal function, so cant pass the correct name to `set_error`. There are two possible solutions I’m aware of. Passing the ufunc name as a string to the internal function,  or passing a status code out of the internal function up to the calling ufunc scalar kernel so that `set_error` can be called there. Right now I prefer the status code approach.

**Comment by ilayn** [comment]
Keeping it strictly to the dynamic allocation, if we allocate beforehand as much as we can before we send things to `ufunc` rabbit hole, and leave the remaining a few with big fat warnings, I think that's already a big accomplishment. We can then try to find out what the good way would be for the remaining errors.

**Comment by WarrenWeckesser** [comment]
Instead of explaining in more detail here what I had in mind in my earlier comment, I put together a proof-of-concept in https://github.com/scipy/scipy/pull/22080.  Just one of the functions in `specfun.h` is fixed, but the same approach should work for all the functions in `specfun.h` that have unchecked allocations.

**Comment by WarrenWeckesser** [comment]
In gh-22080 I maintained the use of the heap.  An alternative (mentioned above at least once) is to declare the arrays on the stack for the cases where the allocation size is fixed.  This would be ideal for both convenience and thread safety.  Most platforms have a stack size limit, so we can't allocate megabytes, but how much is safe?  Is 5kB OK? 20kB?  [This answer](https://stackoverflow.com/a/64085509/1217358) on stackoverflow shows that some systems have very small stack limits, but are those platforms relevant for SciPy?  The answer includes a quote from Bruno Haible from 2009 that "32 KB is more than you can safely allocate on the stack in a multithreaded program".  If we were to allow for those platforms with the smallest stack size limits, we probably wouldn't want to go beyond a few hundred bytes on the stack, so we would stick to using the heap.

**Comment by WarrenWeckesser** [comment]
A quick note: of the 15 functions in `specfun.h` that have unchecked allocations, only two of them allocate arrays whose size depends on the parameters: `kmn()` andf `sdmn()`.  Note, though, that there is a comment in `segv()` that says `TODO: Following array sizes should be decided dynamically`.

**Comment by WarrenWeckesser** [comment]
Earlier I wrote:

> An alternative (mentioned above at least once) is to declare the arrays on the stack for the cases where the allocation size is fixed.

Yet another alternative for these cases is to have fixed size arrays declared outside the functions as [`thread_local`](https://en.cppreference.com/w/cpp/keyword/thread_local).  I'm going to look into this before pushing https://github.com/scipy/scipy/pull/22080 forward.

**Comment by WarrenWeckesser** [comment]
> Yet another alternative for these cases is to have fixed size arrays declared [...] as [thread_local](https://en.cppreference.com/w/cpp/keyword/thread_local).

This works, but in a limited test, I found that this results in a function such as `pro_ang1` taking 2 to 4 times longer when compiled with gcc on Linux.  (I didn't see this performance penalty with clang.)
